src/server/scene/session_locker.cpp

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "session_locker.h"
+#include "mir/frontend/surface_stack.h"
+#include "mir/console_services.h"
+
+#include <algorithm>
+
+namespace ms = mir::scene;
+
+ms::SessionLocker::SessionLocker(
+    std::shared_ptr<mir::ConsoleServices> const& console_services,
+    std::shared_ptr<mf::SurfaceStack> const& surface_stack)
+    : surface_stack{surface_stack}
+{
+    console_services->register_lock_handler(
+        [&] { request_lock(); },
+        [&] { request_unlock(); });
```

Comment:
Not only is this two-phase initialisation still, it may be initialising an object that's about to be dropped on the floor. (Nothing locally ensures that `console_services` isn't released when the constructor returns.